srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,201 @@
+import com.sap.cds.ql.Insert;
 import com.sap.cds.ql.Select;
+import com.sap.cds.ql.Update;
+import com.sap.cds.ql.cqn.CqnAnalyzer;
+import com.sap.cds.reflect.CdsModel;
+import com.sap.cds.services.ErrorStatuses;
+import com.sap.cds.services.ServiceException;
+import com.sap.cds.services.cds.CdsReadEventContext;
 import com.sap.cds.services.cds.CdsService;
+import com.sap.cds.services.draft.DraftService;
 import com.sap.cds.services.handler.EventHandler;
 import com.sap.cds.services.handler.annotations.After;
+import com.sap.cds.services.handler.annotations.Before;
+import com.sap.cds.services.handler.annotations.On;
 import com.sap.cds.services.handler.annotations.ServiceName;
+import com.sap.cds.services.messages.Messages;
 import com.sap.cds.services.persistence.PersistenceService;
 
+import cds.gen.catalogservice.AddReviewContext;
 import cds.gen.catalogservice.Books;
+import cds.gen.catalogservice.Books_;
 import cds.gen.catalogservice.CatalogService_;
+import cds.gen.catalogservice.ReturnCatalogServiceSubmitOrder;
+import cds.gen.catalogservice.Reviews;
+import cds.gen.catalogservice.Reviews_;
+import cds.gen.catalogservice.SubmitOrderContext;
+import cds.gen.reviewservice.ReviewService_;
+import my.bookshop.BookRatingService;
+import my.bookshop.MessageKeys;
 
 /**
  * Custom business logic for the "Catalog Service" (see cat-service.cds)
  *
  * Handles Reading of Books
+ *
  * Adds Discount Message to the Book Title if too much stock is available
+ *
+ * Provides adding book reviews
  */
 @Component
 @ServiceName(CatalogService_.CDS_NAME)
 class CatalogServiceHandler implements EventHandler {
 
 	private final PersistenceService db;
 
-	CatalogServiceHandler(PersistenceService db) {
+	private final DraftService reviewService;
+
+	private final Messages messages;
+
+	private final BookRatingService bookRatingService;
+
+	private final CqnAnalyzer analyzer;
+
+	CatalogServiceHandler(PersistenceService db, @Qualifier(ReviewService_.CDS_NAME) DraftService reviewService,
+			Messages messages, BookRatingService bookRatingService, CdsModel model) {
 		this.db = db;
+		this.reviewService = reviewService;
+		this.messages = messages;
+		this.bookRatingService = bookRatingService;
+		this.analyzer = CqnAnalyzer.create(model);
+	}
+
+	/**
+	 * Invokes some validations before creating a review.
+	 *
+	 * @param context {@link ReviewContext}
+	 */
+	@Before(entity = Books_.CDS_NAME)
+	public void beforeAddReview(AddReviewContext context) {
+		String reviewer = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+
+		Result result = db.run(Select.from(CatalogService_.REVIEWS)
+				.where(review -> review.book_ID().eq(bookId).and(review.reviewer().eq(reviewer))));
+
+		if (result.first().isPresent()) {
+			throw new ServiceException(ErrorStatuses.METHOD_NOT_ALLOWED,
+					"User not allowed to add more than one review for a given book").messageTarget(Reviews_.class,
```

Comment:
Can we add this error message text to our resource bundles and provide a german translation?

Reply:
Done.